Eliminate memory copy in ByteToMessageDecoder whenever possible

normanmaurer opened the pull request:
Motivation:

Currently when there are bytes left in the cumulation buffer we do a byte copy to produce the input buffer for the decode method. This can put quite some overhead on the impl.

normanmaurer commented:
@trustin I just added the cumulator stuff like suggested and only enabled it for SslHandler by default as parsing SSL Packets is really straight forward even with a CompositeByteBuf. I will do more benchmarks to enable it also for others by default but wondered if we can merge this in the meantime and just open a new issue once more benchmarks show it makes sense. 

WDYT ?

trustin commented on codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java:
```diff
@@ -322,4 +377,24 @@ protected void callDecode(ChannelHandlerContext ctx, ByteBuf in, List<Object> ou
         decode(ctx, in, out);
     }
+
+    private static ByteBuf expandCumulation(ByteBufAllocator alloc, ByteBuf cumulation, int readable) {
```
Use package private so that no synthetic method is created.

trustin commented on handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java:
```diff
@@ -474,21 +473,26 @@ public synchronized SSLEngineResult unwrap(
+                || srcsOffset + srcsLength > srcs.length) {
+            throw new IndexOutOfBoundsException(
+                    "offset: " + srcsOffset + ", length: " + srcsLength +
+                            " (expected: offset <= offset + length <= srcs.length (" + srcs.length + "))");
```
Indentation

trustin commented on handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java:
```diff
@@ -474,21 +473,26 @@ public synchronized SSLEngineResult unwrap(
             throw new IndexOutOfBoundsException(
-                    "offset: " + offset + ", length: " + length +
+                    "offset: " + dstsOffset + ", length: " + dstsLength +
                             " (expected: offset <= offset + length <= dsts.length (" + dsts.length + "))");
```
Not a problem of this commit, but please fix the indentation.

trustin commented on handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java:
```diff
@@ -511,23 +515,56 @@ public synchronized SSLEngineResult unwrap(
+        for (int i = srcsOffset; i < srcsEndOffset; i++) {
+            ByteBuffer src = srcs[i];
+            if (src == null) {
+                throw new IllegalArgumentException();
```
Instead of raising an exception, you could break the loop.

normanmaurer replied:
I'm not sure we want to do this... the user gave an offset and a len, and so there should be no null elements in the array. @trustin WDYT ?

trustin replied:
Ah, yeah.  Why don't we just raise NPE with a proper message?

    throw new NullPointerException("srcs[" + i + ']');

normanmaurer replied:
Yep why not :)

trustin commented on handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java:
```diff
@@ -511,23 +515,56 @@ public synchronized SSLEngineResult unwrap(
+            }
+            len += src.remaining();
+            // protect against protocol overflow attack vector
+            if (len > MAX_ENCRYPTED_PACKET_LENGTH) {
```
This check can be done out of the loop.

trustin commented on handler/src/main/java/io/netty/handler/ssl/SslHandler.java:
```diff
@@ -284,6 +284,15 @@ public SslHandler(SSLEngine engine, boolean startTls, Executor delegatedTaskExec
+         * When using {@link OpenSslEngine} we can call {@link OpenSslEngine#unwrap(ByteBuffer[], ByteBuffer[])} and
+         * so remove the need of memory copies for more performance.
+         */
+        setCumulator(engine instanceof OpenSslEngine ? COMPOSITE_CUMULATOR : MERGE_CUMULATOR);
```
We do `engine instanceof OpenSslEngine` three times in this block. Let's just extract a variable.

trustin commented on handler/src/main/java/io/netty/handler/ssl/SslHandler.java:
```diff
@@ -907,9 +914,9 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
             in.skipBytes(totalLength);
-            final ByteBuffer inNetBuf = in.nioBuffer(startOffset, totalLength);
+            final ByteBuf inNetBuf = in.slice(startOffset, totalLength);
             unwrap(ctx, inNetBuf, totalLength);
```
Instead of creating a new slice, please modify `unwrap(ctx, packet, length)` so that it accepts the case where the reader index of the packet is not zero.

trustin commented:
This pull request makes me think we can optimize `SslHandler` even further.  What we do here basically is to apply special execution path when engine is `OpenSslEngine`.  Why don't we just go further and pass memory addresses and lengths? (hint: Use `CompositeByteBuf.internalComponentAt(int)`)

trustin commented:
Let me cherry-pick the first commit. Could you address the comments and file a new pull request for the second change?

normanmaurer commented:
@trustin I think I would prefer to just address your comments, cherry-pick both commits and open another PR for move over to memory address.

normanmaurer commented:
@trustin or is this what you mean ?

trustin commented:
I actually cherry-picked the first commit already.  Please rebase, address the comment, and ping me then.

trustin commented:
@normanmaurer It's actually up to you and I'm fine with your proposal to get this merged first and the do a follow-up pullreq to optimize it further.

trustin commented:
Filed a follow-up issue: #3318

trustin commented on handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java:
```diff
@@ -474,21 +473,26 @@ public synchronized SSLEngineResult unwrap(
+                || srcsOffset + srcsLength > srcs.length) {
+            throw new IndexOutOfBoundsException(
+                "offset: " + srcsOffset + ", length: " + srcsLength +
+                        " (expected: offset <= offset + length <= srcs.length (" + srcs.length + "))");
```
```
"offset: " + srcsOffset + ", length: " + srcsLength +
        " (expected: offset <= offset + length <= srcs.length (" + srcs.length + "))");
```
could be:
```
"offset: " + srcsOffset + ", length: " + srcsLength +
" (expected: offset <= offset + length <= srcs.length (" + srcs.length + "))");
```
according to our updated style. Same for other multi-line string concats.

trustin commented on handler/src/main/java/io/netty/handler/ssl/SslHandler.java:
```diff
@@ -282,8 +282,18 @@ public SslHandler(SSLEngine engine, boolean startTls, Executor delegatedTaskExec
+         * try to use a {@link CompositeByteBuf} at the first place to make it as fast as possible.
+         *
+         * When using {@link OpenSslEngine} we can call {@link OpenSslEngine#unwrap(ByteBuffer[], ByteBuffer[])} and
+         * so remove the need of memory copies for more performance.
```
Punctuation and clarity
```
 * When using JDK {@link SSLEngine}, we use {@link #MERGE_CUMULATOR} because it works only with
 * one {@link ByteBuffer}.
 *
 * When using {@link OpenSslEngine}, we can use {@link #COMPOSITE_CUMULATOR} because it has
 * {@link OpenSslEngine#unwrap(ByteBuffer[], ByteBuffer[])} which works with multiple {@link ByteBuffer}s
 * and which does not need to do extra memory copies.
```

trustin commented on handler/src/main/java/io/netty/handler/ssl/SslHandler.java:
```diff
@@ -1041,25 +1052,82 @@ private void unwrap(
+        int nioBufferCount = in.nioBufferCount();
+        if (engine instanceof OpenSslEngine && nioBufferCount > 1) {
+            // If OpenSslEngine is in use can use a special unwrap method of OpenSslEngine
+            // that allows us to unwrap with multiple ByteBuffers as source.
```
Grammar and punctuation
```
/**
 * If {@link OpenSslEngine} is in use,
 * we can use a special {@link OpenSslEngine#unwrap(ByteBuffer[], ByteBuffer[])} method
 * that accepts multiple {@link ByteBuffer}s without additional memory copies.
 */
```

trustin commented:
Please cherry-pick after addressing the comments.

normanmaurer commented:
@trustin addressed comments and cherry-picked into 4.0 (f46a3d74d0637aeb51326cceccfbee950d9e6838) , 4.1 (1bb818bb59612f1676bb34eb9cde4ea9fb1d3025) and master ()
